### PR TITLE
git-xet client uses HTTPS endpoint

### PR DESCRIPTION
**Pull request by assafvayner**
This PR should not merge before EOD PST of Tuesday 1/23/2024.

- Bump CAS Protocol version.
- Makes git-xet client's cas_client use the HTTPS data plane CAS protocol.
- processes certificate from initiate requests.

### CODE REVIEW

File: rust/cas_client/src/data_transport.rs [hunk 1]
```diff
@@ -54,22 +56,26 @@ enum RetryError {
...
 
     /// Should only be used for non-success errors
     #[error("Status Error: {0}")]
-    Status(http::StatusCode),
+    Status(hyper::StatusCode),
 }
 fn is_status_retriable(err: &RetryError) -> bool {
     match err {
         RetryError::Hyper(_) => true,
+        RetryError::HyperLegacy(_) => true,
```
Comment (port8080) [inline rust/cas_client/src/data_transport.rs:72]: are you sure that tls errors should be marked retryable?

File: rust/cas_client/src/data_transport.rs [hunk 2]
```diff
@@ -102,18 +108,36 @@ impl DataTransport {
     /// creates the DataTransport instance for the H2 connection using
     /// CasConnectionConfig info, and additional port
     pub async fn from_config(cas_connection_config: CasConnectionConfig) -> Result<Self> {
-        debug!(
+        info!(
```
Comment (port8080) [inline rust/cas_client/src/data_transport.rs:111]: nit: make sure that you actually want this level of log in the final commit

File: rust/cas_client/src/data_transport.rs [hunk 3]
```diff
@@ -102,18 +108,36 @@ impl DataTransport {
...
             .http2_initial_stream_window_size(HTTP2_WINDOW_SIZE)
-            .http2_only(true)
-            .build_http();
-
+            .http2_only(true);
+        let connector = if let Some(root_ca) = &cas_connection_config.root_ca {
+            info!("connector with custom root ca");
+            let ca = Certificate::from_pem(root_ca.as_bytes())?;
+            let tls_connector = native_tls::TlsConnector::builder()
+                .danger_accept_invalid_certs(true)
```
Comment (port8080) [inline rust/cas_client/src/data_transport.rs:127]: nit: this is such a scary function name that it deserves a comment explaining why ok

**Review by port8080 — COMMENTED** [review]
1. Can you run the two versions with `export RUST_LOG="hyper=verbose"` and make sure that all the H2 connection negotiation params are unchanged?
2. Can you run the two versions in `rust-lldb` and ensure that the number of threads are unchanged?

File: rust/cas_client/src/data_transport.rs [hunk 4]
```diff
@@ -102,18 +108,36 @@ impl DataTransport {
...
+                .add_root_certificate(ca)
+                .build()?;
+            let tls_connector = tokio_native_tls::TlsConnector::from(tls_connector);
+            let mut sub_connector = HttpConnector::new();
+            sub_connector.enforce_http(false);
+            let mut connector = HttpsConnector::from((sub_connector, tls_connector));
+            connector.https_only(true);
+            connector
+        } else {
+            info!("default connector");
```
Comment (port8080) [inline rust/cas_client/src/data_transport.rs:137]: when would this branch kick in during normal operations? Based on my reading, this is an error path

File: rust/cas_client/src/remote_client.rs [hunk 5]
```diff
@@ -196,18 +220,20 @@ impl RemoteClient {
...
-        let (http_direct, grpc_direct) = self
+        let InitiateResponseEndpoints {
+            h2, put_complete
+        } = self
             .initiate_cas_server_query(prefix, hash, data.len())
             .instrument(debug_span!("remote_client.initiate"))
             .await?;
 
-        debug!("H2 Put initiate response h2 endpoint: {http_direct}, put complete endpoint {grpc_direct}");
+        debug!("H2 Put initiate response h2 endpoint: {}, put complete endpoint {}", h2.endpoint, put_complete.endpoint);
```
Comment (port8080) [inline rust/cas_client/src/remote_client.rs:231]: nit: is there any value in logging the root_ca received too?

File: rust/cas_client/src/data_transport.rs (hunk 3 above)
Comment (jgodlew) [reply rust/cas_client/src/data_transport.rs:127]: I am having trouble understanding why we need this. Shouldn't `add_root_certificate` update the trust store with our root certificate, thereby making any requests to the CAS server valid?

File: rust/cas_client/src/remote_client.rs (hunk 5 above)
Comment (assafvayner) [reply rust/cas_client/src/remote_client.rs:231]: Will add a log.

File: rust/cas_client/src/data_transport.rs (hunk 4 above)
Comment (assafvayner) [reply rust/cas_client/src/data_transport.rs:137]: This would occur if the CAS server does not send the certificate in the initiate response.

In reference to: https://github.com/xetdata/xet-core/pull/240/files#diff-2047c2e47c26df755b6edd8a9a9a04e3b259758f296d6f6748c70c4855e5591cR38, this is the behavior we would want if CAS is behind this latest version, or if we set the CAS_PROTOCOL_VERSION intentionally at 0.2.0, we should still be able to handle responses lacking the root CA.

However this is an artificial requirement I added, I can treat this as an error case instead.

File: rust/cas_client/src/data_transport.rs (hunk 1 above)
Comment (assafvayner) [reply rust/cas_client/src/data_transport.rs:72]: the HyperLegacy errors are not TLS errors, rather when the client makes a request and making the request failed. This replaces a subset of what used to be under the RetryError::Hyper(_) errors.

File: rust/cas_client/src/data_transport.rs (hunk 3 above)
Comment (assafvayner) [reply rust/cas_client/src/data_transport.rs:127]: After some investigation I found the following:

- Setting danger_accept_invalid_certs will make encryption work regardless of specifying the certificate.
- Setting just the certificate (`add_root_certificate`) without setting the danger_accept_invalid_certs will result in a Connect error which is bizarre, telling me I'm using it wrong or it's not actually working.
  - `Error { kind: Connect, source: Some(Error { code: -67843, message: "The certificate was not trusted." }) }`

Comment (assafvayner) [reply rust/cas_client/src/data_transport.rs:127]: switching to using rustls, over native_tls, solves the issue.

File: rust/cas_client/src/data_transport.rs (hunk 4 above)
Comment (port8080) [reply rust/cas_client/src/data_transport.rs:137]: My vote is KISS and handle it as an error

File: rust/cas_client/src/data_transport.rs (hunk 1 above)
Comment (port8080) [reply rust/cas_client/src/data_transport.rs:69]: ack

**Review by jgodlew — APPROVED** [review]
LGTM! :shipit:

**Comment by assafvayner** [comment]
In both cases the configuration settings logged by hyper/h2 display the same values/patterns.
```
WindowUpdate { stream_id: StreamId(0), size_increment: 5177345 }
Settings { flags: (0x0), max_concurrent_streams: 128, initial_window_size: 65536, max_frame_size: 16777215 }
Settings { flags: (0x0), max_concurrent_streams: 200, initial_window_size: 2147418112, max_frame_size: 16384, max_header_list_size: 16777216 }
Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
Settings { flags: (0x0), enable_push: 0, initial_window_size: 2147418112, max_frame_size: 16384 }
```
new after changes:
```
WindowUpdate { stream_id: StreamId(0), size_increment: 5177345 }
Settings { flags: (0x0), max_concurrent_streams: 128, initial_window_size: 65536, max_frame_size: 16777215 }
Settings { flags: (0x0), max_concurrent_streams: 200, initial_window_size: 2147418112, max_frame_size: 16384, max_header_list_size: 16777216 }
Settings { flags: (0x0), enable_push: 0, initial_window_size: 2097152, max_frame_size: 16384 }
Settings { flags: (0x0), enable_push: 0, initial_window_size: 2147418112, max_frame_size: 16384 }
```

tl;dr same exact settings on each http connection across the previous version and the new.

**Comment by assafvayner** [comment]
Tracking the number of tokio tasks in flight with TLS client vs latest main (no TLS) shows a similar amount of threads smudging a 277MB file with no xorbs cached. max of around 25 tasks, with ~20 most of the runtime.
